### DISCUSSION TITLE
METviewer docker unable to read .stat data

### DISCUSSION THREAD

**Discussion by s3dz** [dtcenter/METplus#2215] [Plot Generation] [answered]
Hello METplus team,
I have been getting this error : ERROR:root:*** <class 'KeyError'> in read_data upper *** whenever I try to load Grid_Stat output.

My MET version is 11.0.1, METdataio is 2.0.2 and I'm using metviewer docker from the METviewer v5.0 with repository and following the steps shown in METviewer/docker README file for configuration . 
I was able to work the MODE output and plot a line graph on the web interface however I cannot work the Grid_stat output. 

here's the log 
************************************************************
INFO:root:METdbload Version: __version__ = "2.0.2"
INFO:root:--- *** --- Start METdbLoad --- *** ---
INFO:root:Begin time: 2023-06-14 10:44:15.547060
WARNING:root:!!! Unknown tag: met_version
INFO:root:database name is: mv_met_out
INFO:root:Initial number of files: 39
DEBUG:root:[--- Start read_data ---]
ERROR:root:*** <class 'KeyError'> in read_data upper ***
INFO:root:    >>> Read time: 0:00:00.032979
DEBUG:root:[--- End read_data ---]
DEBUG:root:local_infile is ON
DEBUG:root:[--- Start apply_indexes ---]
INFO:root:--- *** --- Dropping Indexes --- *** ---
INFO:root:    >>> Apply time: 0:00:01.004798
DEBUG:root:[--- End apply_indexes ---]
DEBUG:root:[--- Start write_file_sql ---]
WARNING:root:Duplicate file /data/NCM/output/Grid_Stat/2022111000/grid_stat_WRF_vs_GPM_(*,*)_480000L_20221111_000000V.stat already in data_file
WARNING:root:Duplicate file /data/NCM/output/Grid_Stat/2022111100/grid_stat_WRF_vs_GPM_(*,*)_720000L_20221112_000000V.stat already in data_file
WARNING:root:Duplicate file /data/NCM/output/Grid_Stat/2022111200/grid_stat_WRF_vs_GPM_(*,*)_960000L_20221113_000000V.stat already in data_file
INFO:root:    >>> Write time File: 0:00:00.017008
DEBUG:root:[--- End write_file_sql ---]
DEBUG:root:[--- Start write_metadata_sql ---]
INFO:root:    >>> Write time Metadata: 0:00:00.002264
DEBUG:root:[--- End write_metadata_sql ---]
DEBUG:root:[--- Start apply_indexes ---]
INFO:root:--- *** --- Loading Indexes --- *** ---
INFO:root:    >>> Apply time: 0:00:00.985417
DEBUG:root:[--- End apply_indexes ---]
INFO:root:    >>> Total load time: 0:00:02.054998
INFO:root:End time: 2023-06-14 10:44:17.602825

INFO:root:For Stat Count 0
INFO:root:For Mode CTS Count 0
INFO:root:For Mode Obj Count 0
INFO:root:For Tcst Count 0
INFO:root:For MTD 2D Count 0
INFO:root:For MTD 3D Single Count 0
INFO:root:For MTD 3D Pair Count 0
INFO:root:--- *** --- End METdbLoad --- *** ---
*********************************************************
kindly let me know if I need to send data, xml file ..etc 
Thank you for your support.

**Comment by venitahagerty** [comment]
Hi, I'm Venita, and I do METdataio support. It looks like you have two possible problems, though they may be related. First, the files you are trying to load are already in the database (WARNING:root:Duplicate file). Can you clean out the existing data or load the files into a new database before trying to load these files again?

Second, you are getting a KeyError. There's something in your data that the program doesn't know how to handle, and I apologize for the inconvenience. To track that down, I'll need your XML and data files. If the files can be posted here, great. If not, if you could tar them up and email them to venita.hagerty@noaa.gov, that would allow me to work with them. Thank you.

**Reply by s3dz** [reply]
Hi Venita 
Thank you for your prompted reply.
 I've deleted the data base and created again then ran the command and here the error msg. 
 ************************************
INFO:root:METdbload Version: __version__ = "2.0.2"
INFO:root:--- *** --- Start METdbLoad --- *** ---
INFO:root:Begin time: 2023-06-14 11:17:08.977041
WARNING:root:!!! Unknown tag: met_version
INFO:root:database name is: mv_met_out
INFO:root:Initial number of files: 39
DEBUG:root:[--- Start read_data ---]
ERROR:root:*** <class 'KeyError'> in read_data upper ***
INFO:root:    >>> Read time: 0:00:00.032830
DEBUG:root:[--- End read_data ---]
DEBUG:root:local_infile is ON
DEBUG:root:[--- Start apply_indexes ---]
INFO:root:--- *** --- Dropping Indexes --- *** ---
ERROR:root:*** Index to drop does not exist in run_sql apply_indexes ***
INFO:root:    >>> Apply time: 0:00:00.000558
DEBUG:root:[--- End apply_indexes ---]
DEBUG:root:[--- Start write_file_sql ---]
/usr/local/lib64/python3.6/site-packages/pymysql/cursors.py:170: Warning: (1265, "Data truncated for column 'mod_date' at row 2")
  result = self._query(query)
/usr/local/lib64/python3.6/site-packages/pymysql/cursors.py:170: Warning: (1265, "Data truncated for column 'mod_date' at row 3")
  result = self._query(query)
INFO:root:    >>> Write time File: 0:00:00.009366
DEBUG:root:[--- End write_file_sql ---]
DEBUG:root:[--- Start write_metadata_sql ---]
INFO:root:    >>> Write time Metadata: 0:00:00.003916
DEBUG:root:[--- End write_metadata_sql ---]
DEBUG:root:[--- Start apply_indexes ---]
INFO:root:--- *** --- Loading Indexes --- *** ---
INFO:root:    >>> Apply time: 0:00:00.984019
DEBUG:root:[--- End apply_indexes ---]
INFO:root:    >>> Total load time: 0:00:01.043711
INFO:root:End time: 2023-06-14 11:17:10.021516

INFO:root:For Stat Count 0
INFO:root:For Mode CTS Count 0
INFO:root:For Mode Obj Count 0
INFO:root:For Tcst Count 0
INFO:root:For MTD 2D Count 0
INFO:root:For MTD 3D Single Count 0
INFO:root:For MTD 3D Pair Count 0
INFO:root:--- *** --- End METdbLoad --- *** ---
***************************************************************
I've sent the data and xml file to the email mentioned above.
Regards

**Comment by venitahagerty** [comment, marked as answer]
Thank you for emailing me your data. I was able to load your data with METdataio 2.0.2 on my laptop without a problem, so I'm not sure what's going on with the docker container. 

Sometime after version 2.02 was released, a bug was fixed that had to do with trying to load invalid files. They should just be ignored, but it was causing problems. You could try not having *.txt files in your load directories.

Also, the file names have some unusual characters in them - I haven't seen (\*,\*) in a file name before. It might help to remove those special characters from your file names. An example file name is:

grid_stat_WRF_vs_GPM_(\*,\*)_480000L_20221111_000000V.stat

It looks like the docker file is using Python 3.8, but the error messages in your second log mention Python 3.6, which could also cause a problem:

/usr/local/lib64/python3.6/site-packages/pymysql/cursors.py:170: Warning: (1265, "Data truncated for column 'mod_date' at row 2")

You could run a command in Docker:
    /usr/bin/python -V
to get the python version.

Also, you can look at the content of mv_load.sh script:
    more /METviewer/bin/mv_load.sh 
and see the value of PYTHON_ENV.

**Comment by s3dz** [comment]
Hi Venita
I am so sorry for the late reply and thank you for your detailed reply it was really helpful for me to narrow down the cause of the problem. 

 As you might have guessed, the problem was related to the python version inside the docker. so what I did is that I added a Volume in the docker-compose.yml :
 volumes:
	- ${PYTHON_DOCKER}:/py_docker
where ${PYTHON_DOCKER} where is the path of my python on my system.

Then , I updated the PYTHON_ENV in mv_load.sh 
Finally, I had to install some libraries again but from inside the docker such as pandas.

I am not sure if I needed to all of these steps or not but eventually everything worked out just fine and I am now able to run grid stat xml.

Thank you so much !
